altmetric/explorer/api/client.py

```diff
@@ -0,0 +1,80 @@
+import base64
+import hashlib
+import hmac
+
+import requests
+
+from .query import Query
+from .response import Response
+
+
+def digest(secret, message):
+    '''Calculates a cryptographic digest based on the user's API secret key and
+    the values of some of the parameters as described here:
+
+    https://www.altmetric.com/explorer/documentation/api#authentication
+    '''
+    if message is None:
+        message = ''
+    hmac_sha1 = hmac.new(secret.encode('utf-8'),
+                         message.encode('utf-8'), hashlib.sha1)
+    signature = base64.b64encode(hmac_sha1.digest()).decode('utf-8')
+    signature = hmac_sha1.hexdigest()
+    return signature
+
+
+class Client:
+    """Top level abstraction over the Altmetric Explorer API.
+    """
+
+    def __init__(self, api_endpoint, api_key, api_secret):
+        """Initialises a new Client object
+
+        Args:
+            api_endpoint (string): the url of the explorer api (usually https://www.altmetric.com/explorer/api)
+            api_key (string): your explorer api key
+            api_secret (string): your explorer api secret key
+
+        Raises:
+            ValueError: if the api key or the api secret is None
+
+        Notes:
+            You can find your api key and secret at https://www.altmetric.com/explorer/settings
+        """
+        if api_key == None or api_secret == None:
+            raise ValueError('api_key and api_secret cannot be None')
+
+        self.api_endpoint = api_endpoint
+        self.api_key = api_key
+        self.api_secret = api_secret
+
+    def get(self, path, **args):
```

Comment:
There are no private methods in python unfortunately!

Reply:
Yeah `private` is not really part of Python.  No idea why!

That said, I just remembered that there is a [convention](https://www.datacamp.com/tutorial/python-private-methods-explained) that the names of "private" methods should be prefixed with `__` to highlight to other developers that the method should not be called from outside the class.

In this case, `get` was kept public deliberately for future proofing and also flexibility so you can do things like this without any metaprogramming magic:

```python
for endpoint in SOME_LIST_OF_ENDPOINTS:
  client.get(endpoint)
```

Also, as a general point, I'm thinking of this as a thin and leaky abstraction to give clients a leg-up rather than trying to hide the API from them because I think it's clearer for them and it means it's less likely we'll need to fix bugs or make enhancements later.